indico/modules/core/client/js/session_expiration.js

```diff
@@ -0,0 +1,217 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2023 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import loginURL from 'indico-url:auth.login';
+import sessionExpirationURL from 'indico-url:core.session_expiration';
+import sessionRefreshURL from 'indico-url:core.session_refresh';
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React, {useState, useEffect, useReducer, useRef, useCallback, useMemo} from 'react';
+import ReactDOM from 'react-dom';
+import {Modal, Button} from 'semantic-ui-react';
+
+import {Translate} from 'indico/react/i18n';
+import {handleAxiosError, indicoAxios} from 'indico/utils/axios';
+
+function sessionExpirationReducer(state, action) {
+  switch (action.type) {
+    case 'ACTIVATE_SESSION_TIMER':
+      return {...state, operation: 'active', count: 120};
+    case 'UPDATE_SESSION_EXPIRY':
+      return {...state, sessionExpiry: action.sessionExpiry};
+    case 'START_SESSION_COUNTDOWN':
+      return {...state, operation: 'countdown', count: action.count};
+    case 'UPDATE_SESSION_COUNTDOWN':
+      return {...state, count: action.count};
+    case 'SESSION_EXPIRE':
+      return {...state, operation: 'expired', count: 0};
+    case 'RESET':
+      return {...state, operation: 'reset'};
+    default:
+      return state;
+  }
+}
+
+export default function SessionExpirationManager({sessionExpiration, hardExpiry}) {
+  const [inProgress, setInProgress] = useState(false);
+  const [state, dispatch] = useReducer(sessionExpirationReducer, {
+    operation: 'active',
+    count: 120,
+    sessionExpiry: sessionExpiration,
+  });
+  const sessionTimeout = useRef();
+  const countdownInterval = useRef();
+  const opened = useRef(false);
+  const {operation, count, sessionExpiry} = state;
+  const sessionExpirationChannel = useMemo(() => new BroadcastChannel('session_expiration'), []);
```

Comment:
I *think* it's safe to create the channel on the global level (and then never close it, because closing channels seems to be mostly meant for long-lived cases like webworkers and SPAs).

When I was testing with [this dummy snippet](https://bpa.st/B4LA) it worked pretty fine.

Anyway, I think doing it in here is probably fine as well so as long as everything works fine I would not change it. ;)

Reply:
I took your idea and I'm now creating the channel on a global level and never closing it. It removed the need to have to use memoize for the channel instantiation in react.

Reply:
Btw I added a new useInterval hook which is a copy of the existing useTimeout hook but for the setInterval api